airflow/auth/managers/fab/fab_auth_manager.py

```diff
@@ -87,35 +132,216 @@ def is_logged_in(self) -> bool:
+        """
+        is_global_authorized = self._is_authorized(method=method, resource_type=RESOURCE_DAG, user=user)
+        if is_global_authorized:
+            return True
+
+        if dag_details and dag_details.id:
+            # Check whether the user has permissions to access a specific DAG
+            resource_dag_name = self._resource_name_for_dag(dag_details.id)
+            return self._is_authorized(method=method, resource_type=resource_dag_name, user=user)
+
+        return False
+
+    @staticmethod
+    def _get_fab_action(method: ResourceMethod) -> str:
+        """
+        Convert the method to a FAB action.
+
+        :param method: the method to convert
+
+        :meta private:
+        """
+        if method not in _MAP_METHOD_NAME_TO_FAB_ACTION_NAME:
+            raise AirflowException(f"Unknown method: {method}")
+        return _MAP_METHOD_NAME_TO_FAB_ACTION_NAME[method]
+
+    @staticmethod
+    def _get_fab_resource_type(dag_access_entity: DagAccessEntity):
+        """
+        Convert a DAG access entity to a FAB resource type.
+
+        :param dag_access_entity: the DAG access entity
+
+        :meta private:
+        """
+        if dag_access_entity not in _MAP_DAG_ACCESS_ENTITY_TO_FAB_RESOURCE_TYPE:
+            raise AirflowException(f"Unknown DAG access entity: {dag_access_entity}")
+        return _MAP_DAG_ACCESS_ENTITY_TO_FAB_RESOURCE_TYPE[dag_access_entity]
+
+    def _resource_name_for_dag(self, dag_id: str) -> str:
+        """
+        Returns the FAB resource name for a DAG id.
+
+        :param dag_id: the DAG id
+
+        :meta private:
+        """
+        root_dag_id = self._get_root_dag_id(dag_id)
+        if root_dag_id == RESOURCE_DAG:
+            return root_dag_id
+        if root_dag_id.startswith(RESOURCE_DAG_PREFIX):
+            return root_dag_id
+        return f"{RESOURCE_DAG_PREFIX}{root_dag_id}"
+
+    @staticmethod
+    def _get_user_permissions(user: BaseUser):
+        """
+        Return the user permissions.
+
+        ACTION_CAN_READ and ACTION_CAN_ACCESS_MENU are merged into because they are very similar.
+        We can assume that if a user has permissions to read variables, they also have permissions to access
+        the menu "Variables".
+
+        :param user: the user to get permissions for
+
+        :meta private:
+        """
+        return [
+            (ACTION_CAN_READ if perm[0] == ACTION_CAN_ACCESS_MENU else perm[0], perm[1])
+            for perm in user.perms
+        ]
+
+    def _get_root_dag_id(self, dag_id: str) -> str:
+        """
+        Return the root DAG id in case of sub DAG, return the DAG id otherwise.
+
+        :param dag_id: the DAG id
+
+        :meta private:
+        """
+        if "." in dag_id:
+            dm = (
+                self.security_manager.appbuilder.get_session.query(DagModel.dag_id, DagModel.root_dag_id)
+                .filter(DagModel.dag_id == dag_id)
+                .first()
+            )
```

Comment:
```suggestion
            dm = self.security_manager.appbuilder.get_session.scalar(
                select(DagModel.dag_id, DagModel.root_dag_id)
                .where(DagModel.dag_id == dag_id)
                .limit(1)
            )
```

SQLAlchemy 2 compatibility. Need to import `select` of course.